modules/azure/src/main/clojure/xtdb/azure/file_watch.clj

```diff
@@ -0,0 +1,88 @@
+        {:keys [servicebus-topic-name subscription-name] :as closable-opts} (setup-topic-subscription opts)
+
+        _ (log/info "Initializing filename list from container " container)
+         ;; Init the filename cache with current files
+        _ (file-list-init opts file-name-cache)
+        
+        url-suffix (if prefix (str "/" prefix) "/")
+        base-file-url (str (.getBlobContainerUrl blob-container-client) url-suffix)
+        processor-client (-> (ServiceBusClientBuilder.)
+                             (.fullyQualifiedNamespace (format "%s.servicebus.windows.net" servicebus-namespace))
+                             (.credential azure-credential)
+                             (.processor)
+                             (.topicName servicebus-topic-name)
+                             (.subscriptionName subscription-name)
+                             (.processMessage (reify Consumer
+                                                (accept [_ msg]
+                                                  (let [parsed-msg (json/read-str (.. ^ServiceBusReceivedMessageContext msg getMessage getBody toString) :key-fn keyword)
+                                                        msg-data (:data parsed-msg)
+                                                        event-type (get {"PutBlob" :create "DeleteBlob" :delete} (:api msg-data))
+                                                        file-url (:url msg-data)
+                                                        file (when (string/starts-with? file-url base-file-url)
+                                                              (subs file-url (count base-file-url)))]
+                                                    (log/info (format "Message received, performing %s on file %s" event-type file))
+                                                    (when (and event-type file)
+                                                      (cond
+                                                        (= event-type :create) (.add file-name-cache file)
+                                                        (= event-type :delete) (.remove file-name-cache file)))))))
+                             (.processError (reify Consumer
+                                              (accept [_ msg]
+                                                      (log/error "Error when processing message from service bus queue - " (.getException ^ServiceBusErrorContext msg)))))
+                             (.buildProcessorClient))]
+
+      ;; Start processing messages from the queue
+    (log/info "Watching for filechanges from container " container)
+    (.start processor-client)
+
+      ;; Return all closeable opts from the function
+    (assoc closable-opts :processor-client processor-client)))
+
+(defn watcher-close-fn [{:keys [^ServiceBusAdministrationClient servicebus-admin-client ^ServiceBusProcessorClient processor-client servicebus-topic-name subscription-name]}]
```

Comment:
contrary to my advice from last week, given that you now have something closable this probably _is_ the time for an object :wink: 

In terms of naming, we tend to prefix these functions with `open-` - i.e. `open-file-list-watcher`, which would then return (probably) a `reify AutoCloseable`

Reply:
Agreed - this is much cleaner. May want to do similar for S3, though I'm sure there may be some other refinements/changes to make there anyway so can probably do a followup for that..